Add updated docs for Altis v5 (#105)
humanmade/altis-enhanced-search: docs/cms-query-integration.md

You can also circumvent the use of Elasticsearch for searches by explicitly setting `ep_integrate` to false in the query arguments or by using the `pre_get_posts` action and checking for certain conditions:

```php
add_action( 'pre_get_posts', function ( WP_Query $query ) {
	// Avoid using Elasticsearch for the 'project' post type.
	if ( ! $query->is_search() || $query->get( 'post_type' ) !== 'project' ) {
		return;
	}

	$query->set( 'ep_integrate', false );
}, 20 );
```

## Autosuggest
Because of the way Elasticsearch works it performs better with complete search terms as you might expect to be submitted from a search form.

In some cases you might want to provide a dynamic interface where results are fetched as the user types. The standard method of search will not produce good results until a full word has been typed out so Altis provides a means of analysing and fetching results based on partial search terms.

In the same way `ep_integrate` can be passed to any query class you can pass the parameter `autosuggest`:

```php
$posts = new WP_Query( [

> 	'autosuggest' => true,

Reviewer: Is `ep_integrate` required too? We should state explicitly either way.

Author: Ah yeah, I'll make it clearer that `ep_integrate` defaults to true when the `s` parameter is used.

# CMS Query Integration

The Search module overrides the default search functionality to query the specialized search index in place of a standard MySQL query. This means all default search operations using the CMS search APIs for posts, users and terms such as `WP_Query`, `WP_User_Query`, `WP_Term_Query` and the REST API search endpoints will transparently make use of the search index by default.

> It is possible to control this behaviour and thus use the search index for non search queries that might otherwise be too slow to do in the database such as meta queries.

Reviewer 2: "non search" should be hyphenated, and there should be commas.

```suggestion
It is possible to control this behaviour, and thus use the search index for non-search queries that might otherwise be too slow to do in the database, such as meta queries.
```

Author: The second comma after database reads a bit funny to me, suggesting this instead:
```suggestion
It is possible to control this behaviour and thus use the search index for non-search queries. The reason for doing this is to improve performance when running complex queries, such as meta queries, that would otherwise be too slow to do in the database.
```

Reviewer 2: LGTM

> Each query class accepts an array of parameters. ElasticPress checks for the value of an `ep_integrate` parameter, if this is true the query is performed by Elasticsearch instead of MySQL. If the `s` (search) parameter is present `ep_integrate` will be set to `true` automatically.

Reviewer 2: Second sentence can be broken up more.

```suggestion
Each query class accepts an array of parameters. ElasticPress checks for the value of an `ep_integrate` parameter. If this is true, the query is performed by Elasticsearch instead of MySQL. If the `s` (search) parameter is present `ep_integrate` will be set to `true` automatically.
```

> The following example sets `ep_integrate` to true for a non search query, that is any query where the `s` parameter is not used.

Reviewer 2: Hyphenate "non search", add period, add comma after "that is"

```suggestion
The following example sets `ep_integrate` to true for a non-search query. That is, any query where the `s` parameter is not used.
```

I'm not sure i like the sentence starting with "That is", so perhaps this sentence can be shortened to:

```
The following example sets `ep_integrate` to true for a non-search query where the `s` parameter is not used.
```

or

```
The following example sets `ep_integrate` to true for a query where the `s` parameter is not used.
```

Since this is a "for example", it should probably end with a `:` instead of a period.

Author: I like the first replacement sentence example, good shout. I think it's important to call out "non-search" still

Author: ```suggestion
The following example sets `ep_integrate` to true for a non-search query where the `s` parameter is not used.
```

```php
$posts = new WP_Query( [
	'ep_integrate' => true, // Use Elasticsearch.
	'post_type' => 'project',
	'meta_query' => [
		[
			'key' => 'color',
			'value' => [ 'blue', 'green' ],
			'compare' => 'IN'
		],
	],
] );
```

> Note in the above example you would need a PHP Code Sniffer exception for using the meta query to pass [Automated Code Review](docs://guides/code-review/README.md).

Reviewer 2: Comma after "in the above example." Also, if a PHPCS ignore is required, why not add it to the example (as has been done in other docs)? And then explain why the ignore is there instead.

```suggestion
Note in the above example, you would need a PHP Code Sniffer exception for using the meta query to pass [Automated Code Review](docs://guides/code-review/README.md).
```

Author: Not sure this one is necessary myself but have noticed that I've been a bit lazy with that sentence.
```suggestion
Note that in the above example you would need a PHP Code Sniffer exception for using the meta query to pass [Automated Code Review](docs://guides/code-review/README.md).
```

Author: Good point re. adding the phpcs rule too. I'll let into what that is.

You can also circumvent the use of Elasticsearch for searches by explicitly setting `ep_integrate` to false in the query arguments or by using the `pre_get_posts` action and checking for certain conditions:

```php
add_action( 'pre_get_posts', function ( WP_Query $query ) {
	// Avoid using Elasticsearch for the 'project' post type.
	if ( ! $query->is_search() || $query->get( 'post_type' ) !== 'project' ) {
		return;
	}

	$query->set( 'ep_integrate', false );
}, 20 );
```

## Autosuggest

> Because of the way Elasticsearch works it performs better with complete search terms as you might expect to be submitted from a search form.

Reviewer 2: Comma

However, this opens up questions about "the way Elasticsearch works", e.g. "_how_ does it work?" This sentence assumes pre-existing knowledge or information about "the way Elasticsearch works". Is there relevant documentation that can be linked to that describes this?

```suggestion
Because of the way Elasticsearch works, it performs better with complete search terms as you might expect to be submitted from a search form.
```

Author: 🤔 yeah. To some extent we do expect a bit of knowledge but perhaps we can avoid that wording altogether, eg.
```suggestion
Because of the default way that Elasticsearch analyses text it performs better with complete search terms, as you might expect to be submitted from a search form.
```

Reviewer 2: I like that phrasing better, but I'd remove `default`. Don't forget the comma (in the right place) ;p

```suggestion
Because of the way Elasticsearch analyzes text, it performs better with complete search terms as you might expect to be submitted from a search form.
```

> In some cases you might want to provide a dynamic interface where results are fetched as the user types. The standard method of search will not produce good results until a full word has been typed out so Altis provides a means of analysing and fetching results based on partial search terms.

Reviewer 2: Comma

```suggestion
In some cases, you might want to provide a dynamic interface where results are fetched as the user types. The standard method of search will not produce good results until a full word has been typed out, so Altis provides a means of analysing and fetching results based on partial search terms.
```

Is this the explanation to the above "way that ES works"?

Author: Not quite, the "way that ES works" was referring to how text is analysed, so how sentences are broken up, how words are stemmed or ignored and so on. To enable typeahead functionality we have to add an additional method of text analysis called [ngrams](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-edgengram-tokenfilter.html) that split words up into chunks rather than doing anything more clever.

Author: Don't think a comma after "cases" is needed. At least when I read it in my head I don't pause naturally there. Suggesting:
```suggestion
In some cases you might want to provide a dynamic interface where results are fetched as the user types. The standard method of search will not produce good results until a full word has been typed out, so Altis provides a means of analysing and fetching results based on partial search terms.
```

Reviewer 2: ...says the person who doesn't put a comma after "at least" 😂

I think it's probably fine to leave out the comma there. It works either way, for the most part.

> In the same way as `ep_integrate` can be passed to any query class you can pass the parameter `autosuggest` along with the `s` parameter:

Reviewer 2: Comma

```suggestion
In the same way as `ep_integrate` can be passed to any query class, you can pass the parameter `autosuggest` along with the `s` parameter:
```

```php
$posts = new WP_Query( [
	'autosuggest' => true,
	// A partial search term.
	// Setting `s` will cause `ep_integrate` to be added and set to true.
	's' => 'str',
] );
```

The `autosuggest` parameter defaults to true if the `DOING_AJAX` constant is defined and `true`.

### Filters

> By default only a subset of fields are analysed for autosuggest searches but this can be filtered to add additional fields for each indexable object type.

Reviewer 2: Comma

```suggestion
By default only a subset of fields are analysed for autosuggest searches, but this can be filtered to add additional fields for each indexable object type.
```
